src/main/java/net/starlark/java/eval/Starlark.java

```diff
@@ -681,17 +681,16 @@ public static boolean hasattr(StarlarkSemantics semantics, Object x, String name
    * defaultValue} if non-null, or throws an EvalException otherwise.
    */
   public static Object getattr(
-      Mutability mu,
-      StarlarkSemantics semantics,
+      StarlarkThread thread,
```

Comment:
getattr is an accessor for a data structure. It must not require a thread. Ditto hasattr and dir. It might be acceptable for it to take a DescriptorCache as an optional parameter merely for optimization, but it already has too many parameters.

Optimizations of this small magnitude do not justify making the API more complicated. If that means we need to work harder within the package, so be it.